web/src/search.js

```diff
@@ -165,14 +175,57 @@ export function initialize() {
                 $search_query_box.trigger("blur");
                 update_buttons_with_focus(false);
             }
+
+            if (e.key === "Backspace" && $search_query_box.val() === "") {
+                // We just pressed Backspace and the box was empty. This
+                // dismisses the typeahead, which leaves us in a broken / ugly
+                // UI state due to the box shadow on the search query box. To
+                // prevent this, we trigger a lookup to bring the typeahead
+                // back.
+                open_typeahead();
+            }
         });
 
     // Some of these functions don't actually need to be exported,
     // but the code was moved here from elsewhere, and it would be
     // more work to re-order everything and make them private.
 
-    $search_query_box.on("focus", focus_search);
+    // We don't want to make this a focus handler because selecting the
+    // typehead seems to trigger this (and we don't want to open search
+    // when an option is selected and we're closing search).
+    // Instead we explicitly initiate search on click and on specific keyboard
+    // shortcuts.
+    // TODO(pre-merge): this still doesn't work great for keyboard navigation by tab,
+    // because the typeahead opens on focus but the search bar doesn't get
+    // fully expanded.
+    $search_query_box.on("click", initiate_search);
+
     $search_query_box.on("blur", (e) => {
+        // Clicking on the search icon doesn't count as a blur
+        if ($(e.relatedTarget).closest(".search_button").length) {
+            return;
+        }
+
+        // Don't close the search bar if the user has changed
+        // the text from the default, they might accidentally
+        // click away and not want to lose it.
+        if (message_view_header.get_initial_search_string() !== $("#search_query").val()) {
+            return;
+        }
+
+        if (page_params.search_pills_enabled) {
```

Comment:
TODO: I haven't tested anything to do with pills or the new search, and also don't actually know that much about the development plans around it or who to talk to.

Reply:
#25669 removes that code. It would be great to merge that before this PR gets a thorough review